File: sdks/node/taichi-plugin.js
'use strict';

/**
 * taichi plugin SDK (Node.js implementation, CommonJS).
 *
 * Wraps the stdin/stdout JSON exchange of the taichi plugin protocol so that
 * plugin authors can focus on the test logic. Protocol fields are kept
 * strictly aligned with PluginInput / PluginOutput / PluginCase in
 * taichi/pkg/skill/plugin/skill.go.
 *
 * Protocol overview:
 *   stdin  → PluginInput  JSON (skill_name, project_name, base_url, reports_dir, config)
 *   stdout ← PluginOutput JSON (cases[], error)
 *   stderr ← free-form logs (taichi forwards them to its own logger)
 *   exit 0 = plugin executed normally; exit ≠ 0 = plugin-level fatal error
 *
 * Typical usage:
 *   const { runPlugin, passCase } = require('./taichi-plugin');
 *   runPlugin((input) => {
 *     return { cases: [passCase('Bootstrap')] };
 *   });
 *
 * Uses only Node built-in modules; no third-party dependencies required.
 */

const fs = require('fs');

// Synchronously read all of fd 0 (stdin).
// Node's process.stdin is an async stream by default; here we use
// readFileSync(0) for synchronous reads so the handler can consume it synchronously.
function readStdinSync() {
  try {
    return fs.readFileSync(0, 'utf8');
  } catch (err) {
    // In some pipe scenarios reading may fail; return empty and let JSON.parse report the error upstream.
    return '';
  }
}

/**
 * Read and parse the PluginInput JSON from stdin.
 * @returns {{skill_name:string, project_name:string, base_url:string, reports_dir:string, config:object}}
 */
function readInput() {
  const raw = readStdinSync();
  if (!raw.trim()) {
    throw new Error('plugin input is empty');
  }
  const obj = JSON.parse(raw);
  if (typeof obj !== 'object' || obj === null || Array.isArray(obj)) {
    throw new Error('plugin input must be a JSON object');
  }
  const config =
    obj.config && typeof obj.config === 'object' && !Array.isArray(obj.config)
      ? obj.config
      : {};
  return {
    skill_name: typeof obj.skill_name === 'string' ? obj.skill_name : '',
    project_name: typeof obj.project_name === 'string' ? obj.project_name : '',
    base_url: typeof obj.base_url === 'string' ? obj.base_url : '',
    reports_dir: typeof obj.reports_dir === 'string' ? obj.reports_dir : '',
    config,
  };
}

/**
 * Read the endpoint list from input.config.endpoints; elements are coerced to strings.
 * @param {object} input return value of readInput()
 * @returns {string[]}
 */
function endpoints(input) {
  const raw = input && input.config ? input.config.endpoints : null;
  if (!Array.isArray(raw)) return [];
  return raw.map((item) => String(item));
}

// Normalize a single case into a protocol JSON dict, omitting empty optional fields to align with Go omitempty.
function caseToDict(c) {
  const out = { name: String(c.name), passed: Boolean(c.passed) };
  if (c.skipped) out.skipped = true;
  if (c.message) out.message = String(c.message);
  if (c.duration_ms) out.duration_ms = Number(c.duration_ms);
  if (c.error) out.error = String(c.error);
  return out;
}

/**
 * Serialize a PluginOutput object as JSON and write it to stdout.
 * @param {{cases:Array, error?:string}} output
 * @param {() => void} [done] callback invoked after the write completes (used to flush the buffer before safe exit)
 */
function writeOutput(output, done) {
  const cases = Array.isArray(output.cases)
    ? output.cases.map(caseToDict)
    : [];
  const out = { cases };
  if (output && typeof output.error === 'string' && output.error) {
    out.error = output.error;
  }
  const data = JSON.stringify(out) + '\n';
  if (typeof done === 'function') {
    process.stdout.write(data, () => done());
  } else {
    process.stdout.write(data);
  }
}

/**
 * Construct a passing case.
 * @param {string} name
 * @param {string} [message='ok']
 * @returns {{name:string, passed:boolean, message:string}}
 */
function passCase(name, message = 'ok') {
  return { name, passed: true, message };
}

/**
 * Construct a failing case; `error` is required.
 * @param {string} name
 * @param {string} error
 * @param {string} [message='failed']
 * @returns {{name:string, passed:boolean, message:string, error:string}}
 */
function failCase(name, error, message = 'failed') {
  return { name, passed: false, message, error };
}

/**
 * Construct a skipped case. passed=false, skipped=true.
 * @param {string} name
 * @param {string} [message='skipped']
 * @returns {{name:string, passed:boolean, skipped:boolean, message:string}}
 */
function skipCase(name, message = 'skipped') {
  return { name, passed: false, skipped: true, message };
}

// Format an error thrown by a handler into a protocol error string.
function formatHandlerError(err) {
  const name = err && err.name ? err.name : 'Error';
  const msg = err && err.message ? err.message : String(err);
  return `plugin handler raised: ${name}: ${msg}`;
}

/**
 * Plugin main entry point.
 *
 * Automatically: reads stdin → parses PluginInput → invokes handler →
 * writes PluginOutput to stdout. When the handler throws (or the returned
 * Promise rejects), outputs a PluginOutput with an `error` field and exits
 * with code 1, indicating a plugin-level fatal error.
 *
 * The handler may either synchronously return an output object or return a
 * Promise (async handler). Before exit, stdout is flushed to avoid losing
 * buffered output on the pipe.
 * @param {(input:object) => ({cases:Array, error?:string} | Promise<{cases:Array, error?:string}>)} handler
 */
function runPlugin(handler) {
  let input;
  try {
    input = readInput();
  } catch (err) {
    writeOutput(
      { cases: [], error: `read plugin input failed: ${err.message}` },
      () => process.exit(1)
    );
    return;
  }

  let result;
  try {
    result = handler(input);
  } catch (err) {
    writeOutput(
      { cases: [], error: formatHandlerError(err) },
      () => process.exit(1)
    );
    return;
  }

  // Support async handlers: when a Promise is returned, await its result asynchronously.
  if (result && typeof result.then === 'function') {
    result.then(
      (output) => {
        writeOutput(output || { cases: [] }, () => process.exit(0));
      },
      (err) => {
        writeOutput(
          { cases: [], error: formatHandlerError(err) },
          () => process.exit(1)
        );
      }
    );
    return;
  }

  writeOutput(result || { cases: [] }, () => process.exit(0));
}

module.exports = {
  readInput,
  writeOutput,
  runPlugin,
  endpoints,
  passCase,
  failCase,
  skipCase,
  caseToDict,
};

File: sdks/python/taichi_plugin.py
"""taichi plugin SDK (Python implementation).

Wraps the stdin/stdout JSON exchange of the taichi plugin protocol so that
plugin authors can focus on the test logic itself. Protocol fields are kept
strictly aligned with PluginInput / PluginOutput / PluginCase in
taichi/pkg/skill/plugin/skill.go.

Protocol overview:
  stdin  → PluginInput  JSON (skill_name, project_name, base_url, reports_dir, config)
  stdout ← PluginOutput JSON (cases[], error)
  stderr ← free-form logs (taichi forwards them to its own logger)
  exit 0 = plugin executed normally (pass/fail is expressed via stdout JSON);
          exit ≠ 0 = plugin-level fatal error

Typical usage:

    import taichi_plugin

    def handler(input: taichi_plugin.PluginInput) -> taichi_plugin.PluginOutput:
        cases = [taichi_plugin.pass_case("Bootstrap")]
        return taichi_plugin.PluginOutput(cases=cases)

    if __name__ == "__main__":
        taichi_plugin.run_plugin(handler)

Depends only on the Python standard library; no third-party packages required.
"""

from __future__ import annotations

import json
import sys
from dataclasses import dataclass, field
from typing import Any, Callable, Dict, List

# Type signature of a handler: takes PluginInput and returns PluginOutput.
PluginHandler = Callable[["PluginInput"], "PluginOutput"]


@dataclass
class PluginInput:
    """Input that taichi writes to the plugin's stdin.

    Field names use snake_case and are explicitly mapped to JSON keys via
    from_dict. The `config` field is the passthrough of the `raw` section in
    taichi.yaml after removing command/args/env/workdir/timeout.
    """

    skill_name: str = ""
    project_name: str = ""
    base_url: str = ""
    reports_dir: str = ""
    config: Dict[str, Any] = field(default_factory=dict)

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "PluginInput":
        """Construct PluginInput from a parsed JSON dict; missing fields use defaults."""
        config = data.get("config") or {}
        if not isinstance(config, dict):
            config = {}
        return cls(
            skill_name=data.get("skill_name", "") or "",
            project_name=data.get("project_name", "") or "",
            base_url=data.get("base_url", "") or "",
            reports_dir=data.get("reports_dir", "") or "",
            config=config,
        )

    def endpoints(self) -> List[str]:
        """Convenience accessor for config["endpoints"].

        `endpoints: [/a, /b]` in taichi.yaml is passed through to
        config.endpoints. Returns an empty list when the value is not a list
        or is missing; elements are coerced to strings.
        """
        raw = self.config.get("endpoints")
        if not isinstance(raw, list):
            return []
        return [str(item) for item in raw]


@dataclass
class PluginCase:
    """Result of a single test case.

    `skipped` and `passed` are independent: when `skipped` is true, the taichi
    reporter counts it under skip statistics; otherwise it is counted as
    pass/fail according to `passed`. `skip_case` defaults to passed=False,
    skipped=True.
    """

    name: str
    passed: bool = False
    skipped: bool = False
    message: str = ""
    duration_ms: int = 0
    error: str = ""

    def to_dict(self) -> Dict[str, Any]:
        """Serialize to a protocol JSON dict, omitting empty optional fields to align with Go omitempty."""
        out: Dict[str, Any] = {
            "name": self.name,
            "passed": self.passed,
        }
        if self.skipped:
            out["skipped"] = True
        if self.message:
            out["message"] = self.message
        if self.duration_ms:
            out["duration_ms"] = self.duration_ms
        if self.error:
            out["error"] = self.error
        return out


@dataclass
class PluginOutput:
    """Output that the plugin writes to stdout.

    `cases` is always present (even if empty); a non-empty `error` indicates a
    plugin-level fatal error, in which case taichi marks the skill as not
    fully executed.
    """

    cases: List[PluginCase] = field(default_factory=list)
    error: str = ""

    def to_dict(self) -> Dict[str, Any]:
        out: Dict[str, Any] = {"cases": [c.to_dict() for c in self.cases]}
        if self.error:
            out["error"] = self.error
        return out


def pass_case(name: str, message: str = "ok") -> PluginCase:
    """Construct a passing case."""
    return PluginCase(name=name, passed=True, message=message)


def fail_case(name: str, error: str, message: str = "failed") -> PluginCase:
    """Construct a failing case; `error` is required to describe the failure."""
    return PluginCase(name=name, passed=False, message=message, error=error)


def skip_case(name: str, message: str = "skipped") -> PluginCase:
    """Construct a skipped case. passed=False, skipped=True."""
    return PluginCase(name=name, passed=False, skipped=True, message=message)


def read_input(stream=None) -> PluginInput:
    """Read and parse PluginInput JSON from `stream`; defaults to sys.stdin."""
    stream = stream if stream is not None else sys.stdin
    data = json.load(stream)
    if not isinstance(data, dict):
        raise ValueError("plugin input must be a JSON object")
    return PluginInput.from_dict(data)


def write_output(output: PluginOutput, stream=None) -> None:
    """Serialize PluginOutput as JSON and write it to `stream`; defaults to sys.stdout."""
    stream = stream if stream is not None else sys.stdout
    json.dump(output.to_dict(), stream, ensure_ascii=False)
    stream.write("\n")


def run_plugin(handler: PluginHandler) -> None:
    """Plugin main entry point.

    Automatically: reads stdin → parses PluginInput → invokes handler →
    writes PluginOutput to stdout. When the handler raises an exception,
    outputs a PluginOutput with an `error` field and exits with code 1,
    indicating a plugin-level fatal error.

    Normal execution exits with code 0; pass/fail is expressed via the
    `cases` in stdout JSON.
    """
    try:
        plugin_input = read_input()
    except Exception as exc:  # noqa: BLE001 - protocol-level fallback; input errors are treated as fatal
        write_output(PluginOutput(error=f"read plugin input failed: {exc}"))
        sys.exit(1)

    try:
        output = handler(plugin_input)
    except Exception as exc:  # noqa: BLE001 - convert handler exceptions to plugin-level errors
        write_output(
            PluginOutput(
                error=f"plugin handler raised: {type(exc).__name__}: {exc}"
            )
        )
        sys.exit(1)

    if output is None:
        output = PluginOutput(cases=[])

    write_output(output)
    sys.exit(0)
